README.md
# NAME

Device::L3GD20 - I2C interface to L3GD20 3 axis GyroScope using Device::SMBus

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-brightgreen.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/shantanubhadoria/perl-Device-L3GD20"><img src="https://api.travis-ci.org/shantanubhadoria/perl-Device-L3GD20.svg?branch=build/master" alt="Travis status" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Device-L3GD20%200.010"><img src="https://badgedepot.code301.com/badge/cpantesters/Device-L3GD20/0.010" alt="CPAN Testers result" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Device-L3GD20-0.010"><img src="https://badgedepot.code301.com/badge/kwalitee/Device-L3GD20/0.010" alt="Distribution kwalitee" /></a>
    <a href="https://gratipay.com/shantanubhadoria"><img src="https://img.shields.io/gratipay/shantanubhadoria.svg" alt="Gratipay" /></a>
    </p>
</div>

# VERSION

version 0.010

# ATTRIBUTES

## I2CBusDevicePath

this is the device file path for your I2CBus that the L3GD20 is connected on e.g. /dev/i2c-1
This must be provided during object creation.

## Gyroscope

    $self->Gyroscope->enable();
    $self->Gyroscope->getReading();

This is a object of [Device::Gyroscope::L3GD20](https://metacpan.org/pod/Device::Gyroscope::L3GD20)

# SUPPORT

## Bugs / Feature Requests

Please report any bugs or feature requests through github at 
[https://github.com/shantanubhadoria/perl-device-l3gd20/issues](https://github.com/shantanubhadoria/perl-device-l3gd20/issues).
You will be notified automatically of any progress on your issue.

## Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

[https://github.com/shantanubhadoria/perl-device-l3gd20](https://github.com/shantanubhadoria/perl-device-l3gd20)

    git clone git://github.com/shantanubhadoria/perl-device-l3gd20.git

# AUTHOR

Shantanu Bhadoria &lt;shantanu at cpan dott org>

# CONTRIBUTOR

Shantanu <shantanu@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Changelog for Device-L3GD20

0.010 2016-04-07T06:36:09
 - No changes found

0.009 2015-03-25T07:31:37
 - v0.009 (Shantanu Bhadoria)
 - Pod errors fixed (Shantanu Bhadoria)

0.008 2015-03-25T07:23:19
 - v0.008 (Shantanu Bhadoria)
 - Moved extends, added stopwords for spell checks (Shantanu Bhadoria)

0.007 2013-10-05T00:46:59
 - v0.007 (Shantanu)
 - new readings (Shantanu Bhadoria)

0.006 2013-10-05T00:46:00
 - v0.006 (Shantanu)
 - new readings (Shantanu Bhadoria)
 - Renamed getReading to getReadingDegreesPerSecond (Shantanu Bhadoria)

0.005 2013-09-29T02:46:07
 - v0.005 (Shantanu)
 - Pod Spelling Fixes (Shantanu)

0.004 2013-09-27T10:18:18
 - v0.004 (Shantanu)
 - Xt test fixes (Shantanu)
 - Fixed error in L3GD20.pm (Shantanu)
 - Refactored Libraries to give all similar devices a common prefix for
   easy identification as plugins (Shantanu)
 - minor Changes (Shantanu Bhadoria)

0.003 2013-09-12T09:21:01
 - v0.003 (Shantanu)
 - Added README.pod and Changes (Shantanu)

0.002 2013-09-11T09:13:46
 - pod fix (Shantanu)

0.001 2013-09-11T08:25:26
 - First Release (Shantanu)
 - first commit (Shantanu Bhadoria)
 -  minor commit (Shantanu Bhadoria)
 - First Commit (Shantanu Bhadoria)


examples/example.pl
use FindBin qw($Bin);
use lib "$Bin/../lib";

use Device::Gyroscope::L3GD20;

my $g = Device::Gyroscope::L3GD20->new(
    I2CBusDevicePath => '/dev/i2c-1',
    XZero => -26.99,
    YZero => 7.46,
    ZZero => -9.07,
);

$g->enable();
use Data::Dumper;
my ($x,$y,$z) = (0,0,0);
my ($dx,$dy,$dz) = (0,0,0);
my $count = 0;
while(){
    $gyro = $g->getReading();
    $graw = $g->getRawReading();
#    print 'Gyroscope(Raw): ' . Dumper $g->getRawReading();
#    print "Gyroscope: $gyro->{x}\t$gyro->{y}\t$gyro->{z}\n" ;
    $dx += $gyro->{x}*$g->timeDrift;
    $dy += $gyro->{y}*$g->timeDrift;
    $dz += $gyro->{z}*$g->timeDrift;
    print "Gyroscope: $gyro->{x}\t$gyro->{y}\t$gyro->{z}\n" ;
#    print "Gyroscope(aggregate): $dx\t$dy\t$dz\n" ;
#    print "TIME DRIFT: " . $g->timeDrift . "\n" ;


=head2

    $x+= $graw->{x};
    $y+= $graw->{y};
    $z+= $graw->{z};
    $count++;
    print "AVERAGE: " . ($x/$count) . ", " . ($y/$count) . ", " . ($z/$count) . "\n";

=cut

}


lib/Device/L3GD20/Gyroscope.pm
package Device::L3GD20::Gyroscope;

# PODNAME: Device::L3GD20::Gyroscope
# ABSTRACT: I2C interface to Gyroscope on the L3GD20 using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

use Time::HiRes qw(time);

extends 'Device::SMBus';

=attr I2CDeviceAddress

Containd the I2CDevice Address for the bus on which your gyroscope is connected. It would look like 0x6b. Default is 0x6b.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x6b,
);

=attr gyroscopeGain

Unless you are modifying gyroscope setup you must not change this. This contains the Gyroscope gain value which helps in converting the raw measurements from gyroscope register in to degrees per second.

=cut

has gyroscopeGain => (
    is      => 'rw',
    default => 0.07,
);

=attr XZero
    
This is the raw value for the X axis when the gyro is stationary. This is a part of gyro calibration to get more accurate values for rotation.

=cut

has XZero => (
    is      => 'rw',
    default => 0,
);

=attr YZero
    
This is the raw value for the Y axis when the gyro is stationary. This is a part of gyro calibration to get more accurate values for rotation.

=cut

has YZero => (
    is      => 'rw',
    default => 0,
);

=attr ZZero
    
This is the raw value for the Z axis when the gyro is stationary. This is a part of gyro calibration to get more accurate values for rotation.

=cut

has ZZero => (
    is      => 'rw',
    default => 0,
);

=attr lastReadingTime

Time when the last reading was taken

=cut

has lastReadingTime => (
    is      => 'rw',
    default => 0,
);

=attr timeDrift

time taken between last reading and current reading. 
Make sure to take atleast two readings before you use this value for any calculations in your program otherwise you will get a bad value in this the first time you read the gyroscope.

=cut

has timeDrift => (
    is      => 'rw',
    default => 0,
);

# Registers for the Gyroscope 
use constant {
    CTRL_REG1 => 0x20,
    CTRL_REG4 => 0x23,
};

# X, Y and Z Axis Gyroscope Data value in 2's complement
use constant {
    OUT_X_H => 0x29,
    OUT_X_L => 0x28,

    OUT_Y_H => 0x2b,
    OUT_Y_L => 0x2a,

    OUT_Z_H => 0x2d,
    OUT_Z_L => 0x2c,
};

#use integer; # Use arithmetic right shift instead of unsigned binary right shift with >> 4

=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(CTRL_REG1,0b00001111);
    $self->writeByteData(CTRL_REG4,0b00110000);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers. Note that if XZero,YZero or ZZero are set, this function returns the values adjusted from the values at default non rotating state of the gyroscope. Its recommended that you set these values to achieve accurate results from the gyroscope.

=cut

sub getRawReading {
    my ($self) = @_;

    my $currentTime = time;
    $self->timeDrift( $currentTime - $self->lastReadingTime ) if $self->lastReadingTime;
    $self->lastReadingTime( $currentTime );

    return {
        x => ( $self->_typecast_int_to_int16( $self->readNBytes(OUT_X_L,2) ) ) - $self->XZero,
        y => ( $self->_typecast_int_to_int16( $self->readNBytes(OUT_Y_L,2) ) ) - $self->YZero,
        z => ( $self->_typecast_int_to_int16( $self->readNBytes(OUT_Z_L,2) ) ) - $self->ZZero,
    };
}

=method getReading

Return gyroscope readings in degrees per second

=cut

sub getReading {
    my ($self) = @_;

    my $gain = $self->gyroscopeGain;
    my $gyro = $self->getRawReading;
    return {
        x => ( $gyro->{x} * $gain ),
        y => ( $gyro->{y} * $gain ),
        z => ( $gyro->{z} * $gain ),
    };
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

sub calibrate {
    my ($self) =@_;

}

1;


lib/Device/Gyroscope/L3GD20.pm
use strict;
use warnings;
package Device::Gyroscope::L3GD20;

# PODNAME: Device::Gyroscope::L3GD20
# ABSTRACT: I2C interface to Gyroscope on the L3GD20 using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use POSIX;

use Math::Trig  qw(deg2rad);

use Moose;
extends 'Device::SMBus';

=attr I2CDeviceAddress

Contains the I2CDevice Address for the bus on which your gyroscope is connected. It would look like 0x6b. Default is 0x6b.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x6b,
);

=attr gyroscopeGain

Unless you are modifying gyroscope setup you must not change this. This contains the Gyroscope gain value which helps in converting the raw measurements from gyroscope register in to degrees per second.

=cut

has gyroscopeGain => (
    is      => 'rw',
    default => 0.07,
);

=attr xZero
    
This is the raw value for the X axis when the gyro is stationary. This is a part of gyro calibration to get more accurate values for rotation.

=cut

has xZero => (
    is      => 'rw',
    default => 0,
);

=attr yZero
    
This is the raw value for the Y axis when the gyro is stationary. This is a part of gyro calibration to get more accurate values for rotation.

=cut

has yZero => (
    is      => 'rw',
    default => 0,
);

=attr zZero
    
This is the raw value for the Z axis when the gyro is stationary. This is a part of gyro calibration to get more accurate values for rotation.

=cut

has zZero => (
    is      => 'rw',
    default => 0,
);

=register CTRL_REG1

=register CTRL_REG4

=cut

# Registers for the Gyroscope 
use constant {
    CTRL_REG1 => 0x20,
    CTRL_REG4 => 0x23,
};


=register OUT_X_H

=register OUT_X_L

=register OUT_Y_H

=register OUT_Y_L

=register OUT_Z_H

=register OUT_Z_L

=cut

# X, Y and Z Axis Gyroscope Data value in 2's complement
use constant {
    OUT_X_H => 0x29,
    OUT_X_L => 0x28,

    OUT_Y_H => 0x2b,
    OUT_Y_L => 0x2a,

    OUT_Z_H => 0x2d,
    OUT_Z_L => 0x2c,
};

#use integer; # Use arithmetic right shift instead of unsigned binary right shift with >> 4

=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(CTRL_REG1,0b00001111);
    $self->writeByteData(CTRL_REG4,0b00110000);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from registers. Note that if xZero,yZero or zZero are set, this function returns the values adjusted from the values at default non rotating state of the gyroscope. Its recommended that you set these values to achieve accurate results from the gyroscope.

=cut

sub getRawReading {
    my ($self) = @_;

    return {
        x => ( $self->_typecast_int_to_int16( $self->readNBytes(OUT_X_L,2) ) ) - $self->xZero,
        y => ( $self->_typecast_int_to_int16( $self->readNBytes(OUT_Y_L,2) ) ) - $self->yZero,
        z => ( $self->_typecast_int_to_int16( $self->readNBytes(OUT_Z_L,2) ) ) - $self->zZero,
    };
}

=method getReadingDegreesPerSecond

Return gyroscope readings in degrees per second

=cut

sub getReadingDegreesPerSecond {
    my ($self) = @_;

    my $gain = $self->gyroscopeGain;
    my $gyro = $self->getRawReading;
    return {
        x => ( $gyro->{x} * $gain ),
        y => ( $gyro->{y} * $gain ),
        z => ( $gyro->{z} * $gain ),
    };
}

=method getReadingRadiansPerSecond

Return gyroscope readings in radians per second

=cut

sub getReadingRadiansPerSecond {
    my ($self) = @_;

    my $gain = $self->gyroscopeGain;
    my $gyro = $self->getRawReading;
    return {
        x => deg2rad( $gyro->{x} * $gain ),
        y => deg2rad( $gyro->{y} * $gain ),
        z => deg2rad( $gyro->{z} * $gain ),
    };
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

=method calibrate

Placeholder for documentation on calibration

=cut

sub calibrate {
    my ($self) =@_;

}

1;
